fastestimator/network.py

```diff
@@ -13,51 +13,141 @@
-        self._check_ops()
-        self._check_device()
-
-    def _check_ops(self):
+        self._verify_inputs()
+        self.effective_inputs = dict()
+        self.effective_outputs = dict()
+        self.epoch_ops = []
+
+    def _verify_inputs(self):
         for op in self.ops:
-            self.op_outputs = self.op_outputs.union(set(filter(None, to_list(op.outputs))))
-            self.op_inputs = self.op_inputs.union(set(filter(None, to_list(op.inputs))))
+            if isinstance(op, (RepeatScheduler, EpochScheduler)):
+                for epoch_op in op.get_items():
+                    assert isinstance(epoch_op, TensorOp), "unsupported op format, must provide TensorOp in Network"
+            else:
+                assert isinstance(op, TensorOp), "unsupported op format, must provide TensorOp in Network"
 
-    def _check_device(self):
-        if self.framework == "pytorch":
-            self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
-            if self.device.type == "cuda":
-                for model in self.model_list:
-                    model.to(self.device)
+    def load_epoch(self, mode, epoch):
+        self.epoch_ops = get_current_ops(self.ops, mode, epoch)
 
-    def prepare(self):
+    def unload_epoch(self):
         pass
 
+    def get_loss_keys(self) -> Set[str]:
+        loss_keys = set()
+        for op in self.ops:
+            if isinstance(op, (RepeatScheduler, EpochScheduler)):
+                for epoch_op in op.get_items():
+                    if isinstance(epoch_op, UpdateOp):
+                        loss_keys.update(to_list(epoch_op.inputs))
+            else:
+                if isinstance(op, UpdateOp):
+                    loss_keys.update(to_list(op.inputs))
+        return loss_keys
+
+    def get_effective_input_keys(self, mode, total_epoches) -> Set[str]:
+        input_keys = set()
+        produced_keys = set()
+        for epoch in self.get_signature_epoches(total_epoches):
+            for op in get_current_ops(self.ops, mode, epoch):
+                input_keys.update(set(key for key in to_list(op.inputs) if key not in produced_keys))
+                produced_keys.update(to_list(op.outputs))
+        input_keys -= {None}
+        return input_keys
+
+    def get_all_output_keys(self, mode, total_epoches) -> Set[str]:
+        output_keys = set()
+        for epoch in self.get_signature_epoches(total_epoches):
+            for op in get_current_ops(self.ops, mode, epoch):
+                output_keys.update(to_list(op.outputs))
+        output_keys -= {None}
+        return output_keys
+
+    def get_signature_epoches(self, total_epoches):
+        signature_epoches = {0}
+        epoch_keys = {0}
+        repeat_cycles = {1}
+        for x in self.ops:
+            if isinstance(x, EpochScheduler):
+                epoch_keys.update(x.epoch_dict.keys())
+            elif isinstance(x, RepeatScheduler):
+                repeat_cycles.add(x.cycle_length)
```

Comment:
rather than accessing member variables like cycle_length and epoch_dict, scheduler should perhaps have a function that takes the total epochs and returns the list of signature epochs

Reply:
When `RepeatScheduler` and `EpochScheduler` are used together, `RepeatScheduelr` 's signature epoch is dependent upon `EpochScheduler`. 

Therefore it doesn't make sense to have that function.

Reply:
you'd still reconcile the signature epochs at the top level. or are you saying that we can nest schedulers inside each other?

Reply:
here's an example:
when repeat_list is [a, b] and epoch_dict is {0: c, 5:d}, when combining both scheduler, the signature epochs are: {0, 1, 5, 6}

Now question: what should be the output of repeatScheduler's signature epoch?

Reply:
ah, I see